src/NumericalAlgorithms/Interpolation/LinearInterpolationND.hpp

```diff
@@ -0,0 +1,374 @@
+
+      if constexpr (UniformSpacing) {
+        return linearInterpolation1D_Uniform(
+            A, B, xA, xB, data.inverseSpacing[whichDimension], x1);
+      } else {
+        return linearInterpolation1D_General(A, B, xA, xB, x1);
+      }
+    }
+  };
+
+  // Unpack in dimension!
+  template <typename Data_t, size_t whichDimension, typename CompressedIndex_t,
+            typename X1, typename... Xt>
+  struct RecursiveInterpolationKernelGeneral {
+    static inline double const interpolate(
+        Data_t const& data, std::array<size_t, Dimension> const& indexArray,
+        size_t which_var, Xt const&... xin, X1 const& x1) {
+      auto const xA = data.x[whichDimension][indexArray[whichDimension]];
+      auto const xB = data.x[whichDimension][indexArray[whichDimension] + 1];
+
+      auto const A = RecursiveInterpolationKernelGeneral<
+          Data_t, whichDimension - 1,
+          CompressedIndexRecursion_t<CompressedIndex_t, 0>,
+          Xt...>::interpolate(data, indexArray, which_var, xin...);
+      auto const B = RecursiveInterpolationKernelGeneral<
+          Data_t, whichDimension - 1,
+          CompressedIndexRecursion_t<CompressedIndex_t, 1>,
+          Xt...>::interpolate(data, indexArray, which_var, xin...);
+
+      if constexpr (UniformSpacing) {
+        return linearInterpolation1D_Uniform(
+            A, B, xA, xB, data.inverseSpacing[whichDimension], x1);
+      } else {
+        return linearInterpolation1D_General(A, B, xA, xB, x1);
+      }
+    }
+  };
+
+  // Unpack in dimension!
+  template <typename Data_t, typename CompressedIndex_t, typename X1>
+  struct RecursiveInterpolationKernelGeneral<Data_t, 0, CompressedIndex_t, X1> {
+    static inline double const interpolate(
+        Data_t const& data, std::array<size_t, Dimension> const& indexArray,
+        size_t which_var, X1 const& x1) {
+      size_t constexpr whichDimension = 0;
+
+      auto const xA = data.x[whichDimension][indexArray[whichDimension]];
+      auto const xB = data.x[whichDimension][indexArray[whichDimension] + 1];
+
+      auto const A =
+          data.y[which_var +
+                 NumberOfVariables *
+                     CompressedIndexRecursion_t<CompressedIndex_t, 0>::value(
+                         indexArray, data.numberOfPoints)];
+      auto const B =
+          data.y[which_var +
+                 NumberOfVariables *
+                     CompressedIndexRecursion_t<CompressedIndex_t, 1>::value(
+                         indexArray, data.numberOfPoints)];
+
+      if constexpr (UniformSpacing) {
+        return linearInterpolation1D_Uniform(
+            A, B, xA, xB, data.inverseSpacing[whichDimension], x1);
+      } else {
+        return linearInterpolation1D_General(A, B, xA, xB, x1);
+      }
+    }
+  };
+
+ public:
+  inline int getSize(size_t const whichDimension) {
+    return numberOfPoints[whichDimension];
+  }
+
+  inline void extrapolateHigh(size_t whichDimension, bool value) {
```

Comment:
`void extrapolate_high(const size_t dimension, const bool value)` (I'm not sure what `extrapolate_high/low` means. Is this `allow_extrapolation_above_data`? If so, maybe rename to something more explicit?

Reply:
OK!